demos/PinballPlanning.cpp

```diff
@@ -0,0 +1,326 @@
+    double v2 = motion_state->values[3];
+
+    bool inPaddle = false;
+
+    if ((x1 >= paddleX && x1 <= paddleX + pinballSetup.paddleLength) && (x2 >= paddleY && x2 <= paddleY + pinballSetup.paddleWidth / 2) && v2 < 0)
+        inPaddle = true;    // In jump set if both velocity and position vectors are directed toward top of paddle
+    else if ((x1 >= paddleX && x1 <= paddleX + pinballSetup.paddleLength) && (x2 >= paddleY + pinballSetup.paddleWidth / 2 && x2 <= paddleY + pinballSetup.paddleWidth) && v2 > 0)
+        inPaddle = true;    // In jump set if both velocity and position vectors are directed toward bottom of paddle
+    else if ((x1 >= paddleX && x1 <= paddleX + pinballSetup.paddleLength / 4) && (x2 >= paddleY && x2 <= paddleY + pinballSetup.paddleWidth) && v1 > 0)
+        inPaddle = true;    // In jump set if both velocity and position vectors are directed toward left side of paddle
+    else if ((x1 >= paddleX + (3/4) * pinballSetup.paddleLength && x1 <= paddleX + pinballSetup.paddleLength) && (x2 >= paddleY && x2 <= paddleY + pinballSetup.paddleWidth) && v1 < 0)
+        inPaddle = true;    // In jump set if both velocity and position vectors are directed toward left side of paddle
+
+    return inPaddle;
+}
+
+
+/** \brief Jump set is true whenever the ball is on or below the surface and has a downwards velocity. */
+bool jumpSet(ompl::control::HyRRT::Motion *motion)
+{
+    PinballSetup pinballSetup;
+
+    auto *motion_state = motion->state->as<ompl::base::CompoundState>()->as<ompl::base::RealVectorStateSpace::StateType>(0);
+    double x1 = motion_state->values[0];
+    double v1 = motion_state->values[2];
+
+    for (std::vector<double> paddleCoord : pinballSetup.paddleCoords)    // If ball is in any of the paddles
+    {
+        if (inPaddle(motion, paddleCoord))
+            return true;
+    }
+
+    if ((x1 <= 0 && v1 < 0) || (x1 >= 5 && v1 > 0)) // If ball is in either side wall
+        return true;
+
+    return false;
+}
+
+/** \brief Flow set is true whenever the ball is above the surface or has an upwards velocity. */
+bool flowSet(ompl::control::HyRRT::Motion *motion)
+{
+    return !jumpSet(motion);
+}
+
+/** \brief Unsafe set is true whenever the ball is above 10 units from the ground, to reduce time spent planning. */
+bool unsafeSet(ompl::control::HyRRT::Motion *motion)
+{
+    auto *motion_state = motion->state->as<ompl::base::CompoundState>()->as<ompl::base::RealVectorStateSpace::StateType>(0);
+    double x1 = motion_state->values[0];
+    double x2 = motion_state->values[1];
+
+    if (((x1 >= 0 && x1 < 2) || (x1 > 3 && x1 <= 5)) && x2 <= -10)
+        return true;
+    return false;
+}
+
+void flowODE(const ompl::control::ODESolver::StateType &q, const ompl::control::Control *c,
+             ompl::control::ODESolver::StateType &qdot)
+{
+    (void)c;    // No conrol is applied when a state is in the flow set
+
+    // Retrieve the current orientation of the multicopter.
```

Comment:
Probably not related to multicopter? Also probably here or at the discrete simulator is a good place for a description of the variables and controls (very brief as they are simple)